aiogram/fsm/storage/mongo.py

```diff
@@ -0,0 +1,120 @@
+from typing import Any, Dict, Optional
+
+from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorCollection
+
+from aiogram.fsm.state import State
+from aiogram.fsm.storage.base import (
+    BaseStorage,
+    DefaultKeyBuilder,
+    KeyBuilder,
+    StateType,
+    StorageKey,
+)
+
+
+class MongoStorage(BaseStorage):
+    """
+    MongoDB storage required :code:`motor` package installed (:code:`pip install motor`)
+    """
+
+    def __init__(
+        self,
+        client: AsyncIOMotorClient,
+        key_builder: Optional[KeyBuilder] = None,
+        db_name: str = "aiogram_fsm",
+        states_collection_name: str = "states",
+        data_collection_name: str = "data",
+    ) -> None:
+        """
+        :param client: Instance of AsyncIOMotorClient
+        :param key_builder: builder that helps to convert contextual key to string
+        :param db_name: name of the MongoDB database for FSM
+        :param states_collection_name: name of the collection for storing FSM states.
+        :param data_collection_name: name of the collection for storing additional data.
+        """
+        if key_builder is None:
+            key_builder = DefaultKeyBuilder()
+        self._client = client
+        self._db_name = db_name
+        self._states_collection: AsyncIOMotorCollection = self._client[db_name][
+            states_collection_name
+        ]
+        self._data_collection: AsyncIOMotorCollection = self._client[db_name][data_collection_name]
+        self._key_builder = key_builder
+
+    @classmethod
+    def from_url(
+        cls, url: str, connection_kwargs: Dict[str, Any] = {}, **kwargs: Any
+    ) -> "MongoStorage":
+        """
+        Create an instance of :class:`MongoStorage` with specifying the connection string
+
+        :param url: for example :code:`mongodb://user:password@host:port`
+        :param connection_kwargs: see :code:`motor` docs
+        :param kwargs: arguments to be passed to :class:`MongoStorage`
+        :return: an instance of :class:`MongoStorage`
+        """
+        client = AsyncIOMotorClient(url, **connection_kwargs)
+        return cls(client=client, **kwargs)
+
+    async def close(self) -> None:
+        """Cleanup client resources and disconnect from MongoDB."""
+        self._client.close()
+
+    def resolve_state(self, value: StateType) -> Optional[str]:
+        if value is None:
+            return None
+        elif isinstance(value, State):
+            return value.state
+        return str(value)
+
+    async def set_state(self, key: StorageKey, state: StateType = None) -> None:
+        document_id = self._key_builder.build(key, "state")
```

Comment:
What about using single object key instead of creating multiple documents for the same user?

Just use:
```json5
{
    _id: "fsm:bot_id:user_id:chat_id:destiny", 
    state: state,
    data: {...}
}
```

instead of 
```json5
{
    _id: "fsm:bot_id:user_id:chat_id:destiny:state", 
    state: state,
}
```
and
```json5
{
    _id: "fsm:bot_id:user_id:chat_id:destiny:data",
    data: {...}
}
```

Reply:
Sounds reasonable. I will rewrite it within a couple of days. And we will see how it will look like.

Reply:
I have thought about this proposal more carefully and came to the conclusion that it is better not to combine state and data into one collection. Due to the following reasons:
1. Most likely we don’t want to store documents with empty "state" key and empty "data" key at the same time (with only "_id" key), because it would be a waste of memory. If so, then there would be more code, more logic, more queries to define if document become empty and remove it.
2. It seems like there no way to update data with one query if it is stored as nested object. So there also needed more code and more queries to do it.
So the main advantage of using two separated collections for state and data over using one union collection is more simple code and fewer queries to MongoDB.

Reply:
2. https://www.mongodb.com/community/forums/t/how-to-update-a-nested-document-use-operator-in-mongodb/170700

Reply:
2. Finally find how to do that - https://www.mongodb.com/community/forums/t/upsert-nested-object/204990

Reply:
1. @Olegt0rr, I meant the following difficulty for the case if we store state and data together.
State was created:
`{
    _id: "fsm:bot_id:user_id:chat_id:destiny", 
    state: state
}`
Then data was created:
`{
    _id: "fsm:bot_id:user_id:chat_id:destiny", 
    state: state,
    data: {...}
}`
When state (data) would be removed we need to check if data (state) is presented in document. If data (state) not presented in document we should remove whole document, because no reason to store document with only "_id" key, like this:
`{
    _id: "fsm:bot_id:user_id:chat_id:destiny"
}`
I thought that the implementation of this behavior requires more code and additional logic.
But now i see, that maybe it would be not so complicated as I thought. Okay, I'll give it a try.

Reply:
Done